command/plugins.go

```diff
@@ -53,19 +54,30 @@ func (r *multiVersionProviderResolver) ResolveProviders(
 				continue
 			}
 			if !reqd[name].AcceptsSHA256(digest) {
-				// This generic error message is intended to avoid troubling
-				// users with implementation details. The main useful point
-				// here is that they need to run "terraform init" to
-				// fix this, which is covered by the UI code reporting these
-				// error messages.
-				errs = append(errs, fmt.Errorf("provider.%s: installed but not yet initialized", name))
+				errs = append(errs, fmt.Errorf("provider.%s: checksum mismatch", name))
 				continue
 			}
 
 			client := tfplugin.Client(newest)
 			factories[name] = providerFactory(client)
 		} else {
-			errs = append(errs, fmt.Errorf("provider.%s: no suitable version installed", name))
+			msg := fmt.Sprintf("provider.%s: no suitable version installed", name)
+
+			if req := reqd[name]; req.Versions.String() != "" {
+				foundVersions := []string{}
+				for meta := range r.Available.WithName(name) {
+					foundVersions = append(foundVersions, fmt.Sprintf("%q", meta.Version))
+				}
+
+				found := "none"
+				if len(foundVersions) > 0 {
+					found = strings.Join(foundVersions, ", ")
+				}
+
+				msg += fmt.Sprintf("\n  version requirements: %q\n  versions found: %s", req.Versions, found)
```

Comment:
Again I think just some cosmetic tweaks. Probably subjective, so feel free to ignore if you disagree:

```
* provider.aws: a suitable version is not installed
    version constraint: ~> 0.1
    versions installed: none

Terraform automatically discovers provider requirements from your configuration, including
providers used in child modules. To see the requirements and constraints from each module, 
run "terraform providers".
```

Mainly my goal here is to stick with terminology used in some of the other messages: "constraint" as a singular (even though it's plural in our code, I found singular reads better with "any version" placeholder :woman_shrugging:), "versions _installed_" rather than "_found_" to match with the error heading and with other messages.

I also think we should include a note here about how we arrived at this list of requirements. In some of the other messages I specifically called out "terraform providers" in an attempt to answer the inevitable question of "where is this constraint coming from?" when using lots of external modules. (Ideally we should print this message only once rather than repeating it for each missing provider, so probably this should live somewhere else in the code than where I've left this comment...)

Otherwise, some cosmetic/writing style tweaks, which probably just reflect my personal taste rather than being objectively better.

We should also be sure that the version constraint reads `any version` when `req.Versions.Unconstrained()`, for the same reason we did on the other messages. (Empty quotes look like broken behavior.)